File: lib/solver/ffi_conversions.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'package:fraction/fraction.dart';

import '../models/industry_type.dart';
import 'ffi_types.dart';
import 'problem.dart';
import 'schedule.dart';

Fraction ffi2dart_fraction(FfiFraction frac) => Fraction(frac.numerator, frac.denominator);

BatchItem ffi2dart_batchItem(batchItem item) {
  final time = ffi2dart_fraction(item.time);
  return BatchItem(item.runs, item.slots, time);
}

Batch ffi2dart_batch(batch element) {
  final result = Batch();
  for (int i = 0; i < element.size; ++i) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    result[entry.key] = ffi2dart_batchItem(entry.value);
  }
  return result;
}

List<Batch> ffi2dart_batchlist(batchList element) {
  final result = <Batch>[];
  for (int i = 0; i < element.size; ++i) {
    result.add(ffi2dart_batch(element.entries
        .elementAt(i)
        .ref));
  }
  return result;
}

Map<IndustryType, List<Batch>> ffi2dart_machine2batches(k2batches element) {
  final result = <IndustryType, List<Batch>>{};
  for (int i = 0; i < element.size; ++i) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    final key = entry.key == 1 ? IndustryType.MANUFACTURING : IndustryType.REACTION;
    result[key] = ffi2dart_batchlist(entry.value);
  }
  return result;
}

// We use this to parse schedules passed to us from c++
Schedule ffi2dart_schedule(FfiSchedule schedule) {
  Schedule result = Schedule(ffi2dart_machine2batches(schedule.machine2batches));
  result.time = schedule.time;
  result.isOptimized = true;
  return result;
}

void make_fraction(FfiFraction frac, Fraction data) {
  frac.numerator = data.numerator;
  frac.denominator = data.denominator;
}

void make_batchItem(batchItem item, BatchItem data) {
  make_fraction(item.time, data.time);
  item.slots = data.slots;
  item.runs = data.runs;
}

void make_batch(batch b, Batch data) {
  b.size = data.items.length;
  b.entries = calloc.allocate(sizeOf<i2batchItemEntry>() * data.items.length);
  int i = 0;
  for (int key in data.items.keys) {
    final entry = b.entries
        .elementAt(i)
        .ref;
    entry.key = key;
    make_batchItem(entry.value, data.items[key]!);
    i += 1;
  }
}

void make_batchList(batchList element, List<Batch> data) {
  element.size = data.length;
  element.entries = calloc.allocate(sizeOf<batch>() * data.length);
  for (int i = 0; i < data.length; ++i) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    make_batch(entry, data[i]);
  }
}

void make_k2batches(k2batches element, Map<IndustryType, List<Batch>> data) {
  element.size = data.length;
  element.entries = calloc.allocate(sizeOf<k2batchesEntry>() * data.length);
  int i = 0;
  for (final machine in data.keys) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    entry.key = machine == IndustryType.MANUFACTURING ? 1 : 0;
    make_batchList(entry.value, data[machine]!);
    i += 1;
  }
}

Pointer<FfiSchedule> make_schedule(Schedule schedule) {
  final result = calloc<FfiSchedule>();
  result.ref.time = schedule.time;
  make_k2batches(result.ref.machine2batches, schedule.getBatches());
  return result;
}

void destroy_batch(batch b) => calloc.free(b.entries);

void destroy_batchList(batchList element) {
  for (int i = 0; i < element.size; ++i) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    destroy_batch(entry);
  }
  calloc.free(element.entries);
}

void destroy_k2batches(k2batches element) {
  for (int i = 0; i < element.size; ++i) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    destroy_batchList(entry.value);
  }
  calloc.free(element.entries);
}

void destroy_schedule(Pointer<FfiSchedule> schedule) {
  destroy_k2batches(schedule.ref.machine2batches);
  calloc.free(schedule);
}

void make_i2frac(i2frac element, Map<int, Fraction> data) {
  element.size = data.length;
  element.entries = calloc.allocate(sizeOf<i2fracEntry>() * data.length);
  int i = 0;
  for (var key in data.keys) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    entry.key = key;
    make_fraction(entry.value, data[key]!);
    i += 1;
  }
}

void make_i2i(i2i element, Map<int, int> data) {
  element.size = data.length;
  element.entries = calloc.allocate(sizeOf<i2iEntry>() * data.length);
  int i = 0;
  for (var key in data.keys) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    entry.key = key;
    entry.value = data[key]!;
    i += 1;
  }
}

void make_i2i2i(i2i2i element, Map<int, Map<int, int>> data) {
  element.size = data.length;
  element.entries = calloc.allocate(sizeOf<i2i2iEntry>() * data.length);
  int i = 0;
  for (var key in data.keys) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    entry.key = key;
    make_i2i(entry.value, data[key]!);
    i += 1;
  }
}

Pointer<Int32> make_ilist(List<int> data) {
  final result = calloc.allocate(sizeOf<Int32>() * data.length).cast<Int32>();
  for (int i = 0; i < data.length; ++i) {
    result
        .elementAt(i)
        .value = data[i];
  }
  return result;
}

void destroy_i2frac(i2frac element) => calloc.free(element.entries);

void destroy_i2i(i2i element) => calloc.free(element.entries);

void destroy_i2i2i(i2i2i element) {
  for (int i = 0; i < element.size; ++i) {
    final entry = element.entries
        .elementAt(i)
        .ref;
    destroy_i2i(entry.value);
  }
  calloc.free(element.entries);
}

void destroy_ilist(Pointer<Int32> ptr) {
  calloc.free(ptr);
}

Pointer<FfiProblem> make_problem(Problem problem) {
  final result = calloc<FfiProblem>();
  make_i2i(result.ref.timePerRun, problem.timePerRun);
  make_i2i(result.ref.madePerRun, problem.madePerRun);
  make_i2i(result.ref.inventory, problem.inventory.getQuantities());
  make_i2i(result.ref.maxNumRunsPerSlotOfJob, problem.maxNumRunsPerSlotOfJob);
  make_i2i(
      result.ref.maxNumSlotsOfMachine,
      problem.maxNumSlotsOfMachine
          .map((k, v) => MapEntry(k == IndustryType.MANUFACTURING ? 1 : 0, v)));
  make_i2i(result.ref.maxNumSlotsOfJob, problem.maxNumSlotsOfJob);
  make_i2i(result.ref.runsExcess, problem.runsExcess);
  make_i2i(result.ref.job2machine,
      problem.job2machine.map((k, v) => MapEntry(k, v == IndustryType.MANUFACTURING ? 1 : 0)));
  result.ref.tids = make_ilist(problem.tids.toList());
  result.ref.float2int = problem.float2int;
  make_i2i2i(result.ref.dependencies, problem.dependencies);
  if (problem.approximation != null) {
    result.ref.approximation = make_schedule(problem.approximation!);
  }
  make_i2frac(result.ref.materialBonus, problem.jobMaterialBonus);
  make_i2frac(result.ref.timeBonus, problem.jobTimeBonus);
  return result;
}

void destroy_problem(Pointer<FfiProblem> p) {
  destroy_i2i(p.ref.timePerRun);
  destroy_i2i(p.ref.madePerRun);
  destroy_i2i(p.ref.inventory);
  destroy_i2i(p.ref.maxNumRunsPerSlotOfJob);
  destroy_i2i(p.ref.maxNumSlotsOfMachine);
  destroy_i2i(p.ref.maxNumSlotsOfJob);
  destroy_i2i(p.ref.runsExcess);
  destroy_i2i(p.ref.job2machine);
  destroy_ilist(p.ref.tids);
  destroy_i2i2i(p.ref.dependencies);
  if (p.ref.approximation.address != 0) {
    destroy_schedule(p.ref.approximation);
  }
  destroy_i2frac(p.ref.materialBonus);
  destroy_i2frac(p.ref.timeBonus);
  calloc.free(p);
}

File: advanced_solver_cpp/ffi_conversions.cpp
#include "ffi_types.h"
#include "problem.h"
#include "schedule.h"

// we are passed a schedule & problem thru ffi and need to convert it
Fraction ffi2cpp_fraction(FfiFraction frac) { return Fraction(frac.numerator, frac.denominator); }

BatchItem ffi2cpp_batchItem(batchItem& item) {
  auto time = ffi2cpp_fraction(item.time);
  return BatchItem(item.runs, item.slots, time);
}

Batch ffi2cpp_batch(batch& element) {
  auto result = Batch();
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    result.items[entry.key] = ffi2cpp_batchItem(entry.value);
  }
  return result;
}

vector<Batch> ffi2cpp_batchlist(batchList element) {
  auto result = vector<Batch>{};
  for (int i = 0; i < element.size; ++i) {
    result.push_back(ffi2cpp_batch(element.entries[i]));
  }
  return result;
}

map<IndustryType, vector<Batch>> ffi2cpp_machine2batches(k2batches& element) {
  auto result = map<IndustryType, vector<Batch>>{};
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    auto key = entry.key == MachineType::MANUFACTURING_ ? IndustryType::MANUFACTURING : IndustryType::REACTION;
    result[key] = ffi2cpp_batchlist(entry.value);
  }
  return result;
}

Schedule ffi2cpp_schedule(FfiSchedule schedule) {
  Schedule result = Schedule();
  result.machine2batches = ffi2cpp_machine2batches(schedule.machine2batches);
  result.time = schedule.time;
  return result;
}

map<int, int64_t> ffi2cpp_i2i(i2i& element) {
  auto result = map<int, int64_t>{};
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    result[entry.key] = entry.value;
  }
  return result;
}

map<int, map<int, int64_t>> ffi2cpp_i2i2i(i2i2i& element) {
  auto result = map<int, map<int, int64_t>>{};
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    result[entry.key] = ffi2cpp_i2i(entry.value);
  }
  return result;
}

map<int, Fraction> ffi2cpp_i2frac(i2frac& element) {
  auto result = map<int, Fraction>{};
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    result[entry.key] = ffi2cpp_fraction(entry.value);
  }
  return result;
}

map<int, IndustryType> ffi2cpp_i2indy(i2i& element) {
  auto result = map<int, IndustryType>{};
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    result[entry.key] = entry.value == 0 ? IndustryType::REACTION : IndustryType::MANUFACTURING;
  }
  return result;
}

map<IndustryType, int64_t> ffi2cpp_indy2i(i2i& element) {
  auto result = map<IndustryType, int64_t>{};
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    result[entry.key == 0 ? IndustryType::REACTION : IndustryType::MANUFACTURING] = entry.value;
  }
  return result;
}

Problem ffi2cpp_problem(struct FfiProblem p) {
  // clang-format off
  return Problem(ffi2cpp_i2i(p.runsExcess),
                 ffi2cpp_i2i(p.madePerRun),
                 ffi2cpp_i2i(p.timePerRun),
                 ffi2cpp_i2indy(p.job2machine),
                 ffi2cpp_i2i2i(p.dependencies),
                 ffi2cpp_i2i(p.inventory),
                 ffi2cpp_indy2i(p.maxNumSlotsOfMachine),
                 ffi2cpp_i2i(p.maxNumSlotsOfJob),
                 ffi2cpp_i2i(p.maxNumRunsPerSlotOfJob),
                 ffi2cpp_i2frac(p.materialBonus),
                 ffi2cpp_i2frac(p.timeBonus),
                 ffi2cpp_schedule(*p.approximation),
                 p.float2int);
  // clang-format on
}

void make_fraction(FfiFraction& frac, Fraction data) {
  // TODO now that dart uses int64_t not sure if i need this check here!!!!
  if (data.num > 2147483647 || data.den > 2147483647) {
    std::cout << "overflow in make_fraction" << std::endl;
    exit(1);
  }
  frac.numerator = data.num;
  frac.denominator = data.den;
}

void make_batchItem(batchItem& item, BatchItem data) {
  make_fraction(item.time, data.time);
  item.slots = data.slots;
  item.runs = data.runs;
}

void make_batch(batch& b, Batch data) {
  b.startTime = data.startTime;
  b.size = int(data.items.size());
  b.entries = (i2batchItemEntry*)calloc(data.items.size(), sizeof(i2batchItemEntry));
  int i = 0;
  for (auto& [k, v] : data.items) {
    auto& entry = b.entries[i];
    entry.key = k;
    make_batchItem(entry.value, v);
    i += 1;
  }
}

void make_batchList(batchList& element, vector<Batch> data) {
  element.size = int(data.size());
  element.entries = (batch*)calloc(data.size(), sizeof(batch));
  for (int i = 0; i < data.size(); ++i) {
    auto& entry = element.entries[i];
    make_batch(entry, data[i]);
  }
}

void make_k2batches(k2batches& element, map<IndustryType, vector<Batch>> data) {
  element.size = int(data.size());
  element.entries = (k2batchesEntry*)calloc(data.size(), sizeof(k2batchesEntry));
  int i = 0;
  for (auto& [k, v] : data) {
    auto& entry = element.entries[i];
    entry.key = k == IndustryType::MANUFACTURING ? MachineType::MANUFACTURING_ : MachineType::REACTION_;
    make_batchList(entry.value, v);
    i += 1;
  }
}

// we pass a schedule back to dart through the ffi and need to convert it
FfiSchedule* make_schedule(Schedule schedule) {
  auto result = (FfiSchedule*)calloc(1, sizeof(FfiSchedule));
  result->time = schedule.time;
  result->infeasible = schedule.infeasible;
  result->optimal = schedule.optimal;
  make_k2batches(result->machine2batches, schedule.machine2batches);
  return result;
}

void destroy_batch(batch& b) { free(b.entries); }

void destroy_batchList(batchList& element) {
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    destroy_batch(entry);
  }
  free(element.entries);
}

void destroy_k2batches(k2batches& element) {
  for (int i = 0; i < element.size; ++i) {
    auto& entry = element.entries[i];
    destroy_batchList(entry.value);
  }
  free(element.entries);
}

// we make a schedule so we will have to delete the schedule when done with it
void destroy_schedule(FfiSchedule* schedule) {
  destroy_k2batches(schedule->machine2batches);
  free(schedule);
}
